### PR TITLE
Add support for updating vSphere credentials

### PR DESCRIPTION
**Pull request by abhay-krishna**
Fixes https://github.com/aws/eks-anywhere/issues/6468

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/vsphere/vsphere.go [hunk 1]
```diff
@@ -946,16 +956,16 @@ func (p *vsphereProvider) GenerateCAPISpecForCreate(ctx context.Context, _ *type
 	return controlPlaneSpec, workersSpec, nil
 }
 
-func (p *vsphereProvider) createSecret(ctx context.Context, cluster *types.Cluster, contents *bytes.Buffer) error {
+func (p *vsphereProvider) updateAllVsphereSecrets(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec, contents *bytes.Buffer) error {
```
Comment (g-gaston) [inline pkg/providers/vsphere/vsphere.go:959]: this method uses 3 different patterns for k8s resource generation: go templates, go string formatting and go struct yaml marshaling.

Can we just pick one and use it for all secrets? I'm in favor of only using go struct marshaling.

File: pkg/providers/vsphere/vsphere.go [hunk 2]
```diff
@@ -967,11 +977,69 @@ func (p *vsphereProvider) createSecret(ctx context.Context, cluster *types.Clust
...
+		if err != nil {
+			return fmt.Errorf("marshalling secret object: %v", err)
+		}
+
+		err = p.providerKubectlClient.ApplyKubeSpecFromBytes(ctx, cluster, objBytes)
+		if err != nil {
+			return fmt.Errorf("applying secret object: %v", err)
+		}
+	}
+
```
Comment (g-gaston) [inline pkg/providers/vsphere/vsphere.go:1037]: Tests for this new code?

**Review by g-gaston — COMMENTED** [review]
Looking at the issue, it's mentioned that we want a better user experience and that how we handle credentials in vsphere might need a "redesign". I understand this PR doesn't do that and just adds code to automatically update secrets when the user passes new values to the CLI. Is there a long term vision for this feature?

Regarding the change itself, I think there might be some gotchas/nuances to update secrets that are shared across clusters. Could we add in the PR description more information about how this is supposed to behave for different scenarios (management vs workload clusters)? Another way to do this would be to just write down the expected steps a customer needs to take to update the credentials across all clusters.

Btw I'm not saying the change is a bad one, just making sure we build enough context around it. Right now it requires a lot of previous knowledge and inference from the code to understand the implications.

**Comment by abhay-krishna** [comment]
> Looking at the issue, it's mentioned that we want a better user experience and that how we handle credentials in vsphere might need a "redesign". I understand this PR doesn't do that and just adds code to automatically update secrets when the user passes new values to the CLI. Is there a long term vision for this feature?
> 
> Regarding the change itself, I think there might be some gotchas/nuances to update secrets that are shared across clusters. Could we add in the PR description more information about how this is supposed to behave for different scenarios (management vs workload clusters)? Another way to do this would be to just write down the expected steps a customer needs to take to update the credentials across all clusters.
> 
> Btw I'm not saying the change is a bad one, just making sure we build enough context around it. Right now it requires a lot of previous knowledge and inference from the code to understand the implications.

In the long term, I had envisioned adding a command to the CLI that could update the vSphere credentials. We already have the `eksctl anywhere exp vsphere` subcommand for experimental features such as setting up vSphere users. I was thinking about adding a utility command such as `update-password` to the CLI, which customers can leverage to update their vSphere credentials.

I will update the description with the user experience and add documentation in the vSphere docs for this feature.

**Pull request closed without merging**